rasa/telemetry.py

```diff
@@ -0,0 +1,553 @@
+
+
+def initialize_telemetry() -> bool:
+    """Read telemetry configuration from the user's Rasa config file in $HOME.
+
+    Creates a default configuration if no configuration exists.
+
+    Returns:
+        `True`, if telemetry is enabled, `False` otherwise.
+    """
+
+    # calling this even if the environment variable is set makes sure the
+    # configuration is created and there is a telemetry ID
+    is_enabled_in_configuration = _is_telemetry_enabled_in_configuration()
+
+    telemetry_environ = os.environ.get(TELEMETRY_ENABLED_ENVIRONMENT_VARIABLE)
+
+    if telemetry_environ is None:
+        return is_enabled_in_configuration
+    else:
+        return telemetry_environ.lower() == "true"
+
+
+def ensure_telemetry_enabled(f: Callable[..., Any]) -> Callable[..., Any]:
+    """Function decorator for telemetry functions that only runs the decorated
+    function if telemetry is enabled."""
+
+    is_telemetry_enabled = initialize_telemetry()
+
+    # allows us to use the decorator for async and non async functions
+    if asyncio.iscoroutinefunction(f):
+
+        @wraps(f)
+        async def decorated(*args, **kwargs):
+            try:
+                if is_telemetry_enabled:
+                    return await f(*args, **kwargs)
+            except Exception as e:  # skipcq:PYL-W0703
+                logger.debug(f"Skipping telemetry reporting: {e}")
+            return None
+
+        return decorated
+    else:
+
+        @wraps(f)
+        def decorated(*args, **kwargs):
+            try:
+                if is_telemetry_enabled:
+                    return f(*args, **kwargs)
+            except Exception as e:  # skipcq:PYL-W0703
+                logger.debug(f"Skipping telemetry reporting: {e}")
+            return None
+
+        return decorated
+
+
+def telemetry_write_key() -> Optional[Text]:
+    """Read the Segment write key from the segment key text file.
+    The segment key text file should by present only in wheel/sdist packaged
+    versions of Rasa Open Source. This avoids running telemetry locally when
+    developing on Rasa or when running CI builds.
+
+    In local development, this should always return `None` to avoid logging telemetry.
+
+    Returns:
+        Segment write key, if the key file was present.
+    """
+    import pkg_resources
+    from rasa import __name__ as name
+
+    if os.environ.get(TELEMETRY_WRITE_KEY_ENVIRONMENT_VARIABLE):
+        # a write key set using the environment variable will always
+        # overwrite any key provided as part of the package (`segment_key` file)
+        return os.environ.get(TELEMETRY_WRITE_KEY_ENVIRONMENT_VARIABLE)
+
+    write_key_path = pkg_resources.resource_filename(name, "segment_key")
+
+    # noinspection PyBroadException
+    try:
+        with open(write_key_path) as f:
+            return f.read().strip()
+    except Exception:  # skipcq:PYL-W0703
+        return None
+
+
+def encode_base64(original: Text, encoding: Text = "utf-8") -> Text:
```

Comment:
```suggestion
def _encode_base64(original: Text, encoding: Text = "utf-8") -> Text:
```

Reply:
I do want to test it though which means it should not be protected.

Reply:
I wouldn't change production code for the sake of testing code. 
I'd either just call it anyway from the tests or test it using a public function which calls that.

Reply:
it isn't specifically tested anyways now (which makes sense as it shouldn't be a public function) so made it protected.